Mathlib/SetTheory/Descriptive/Tree.lean

```diff
@@ -27,20 +27,96 @@ def tree (A : Type*) : CompleteSublattice (Set (List A)) :=
     (by rintro S hS x a ⟨t, ht, hx⟩; use t, ht, hS ht hx)
     (by rintro S hS x a h T hT; exact hS hT <| h T hT)
 
-instance (A : Type*) : SetLike (tree A) (List A) := SetLike.instSubtypeSet
+@[simps!] instance (A : Type*) : SetLike (tree A) (List A) := SetLike.instSubtypeSet
 
+namespace Tree
 variable {A : Type*} {T : tree A}
 
 lemma mem_of_append {x y : List A} (h : x ++ y ∈ T) : x ∈ T := by
   induction' y with y ys ih generalizing x
   · simpa using h
   · exact T.prop (ih (by simpa))
-
 lemma mem_of_prefix {x y : List A} (h' : x <+: y) (h : y ∈ T) : x ∈ T := by
   obtain ⟨_, rfl⟩ := h'; exact mem_of_append h
+instance : Trans (List.IsPrefix) (fun x (T : tree A) ↦ x ∈ T) (fun x T ↦ x ∈ T) where
+  trans := mem_of_prefix
+lemma singleton_mem (T : tree A) {a : A} {x : List A} (h : a :: x ∈ T) : [a] ∈ T :=
+  mem_of_prefix ⟨x, rfl⟩ h
 
 @[simp] lemma tree_eq_bot : T = ⊥ ↔ [] ∉ T where
   mp := by rintro rfl; simp
   mpr h := by ext x; simpa using fun h' ↦ h <| mem_of_prefix x.nil_prefix h'
 
-end Descriptive
+lemma take_mem {T : tree A} {n : ℕ} (x : T) : x.val.take n ∈ T :=
+  mem_of_prefix (x.val.take_prefix n) x.prop
+
+/-- A variant of `List.take` internally to a tree -/
+@[simps] def take {T : tree A} (n : ℕ) (x : T) : T := ⟨x.val.take n, take_mem x⟩
+
+@[simp] lemma take_take {T : tree A} (m n : ℕ) (x : T) :
+  take m (take n x) = take (m ⊓ n) x := by simp [Subtype.ext_iff, List.take_take]
+@[simp] lemma take_eq_take {x : T} {m n : ℕ} :
+  take m x = take n x ↔ m ⊓ x.val.length = n ⊓ x.val.length := by simp [Subtype.ext_iff]
+
+variable (T) (x y : List A)
+/-- The residual tree obtained by regarding the node x as new root -/
+def subAt : tree A := ⟨(x ++ ·)⁻¹' T, fun _ _ _ ↦ mem_of_append (by rwa [List.append_assoc])⟩
+@[simp] lemma mem_subAt : y ∈ subAt T x ↔ x ++ y ∈ T := Iff.rfl
+
+@[simp] lemma subAt_nil : subAt T [] = T := rfl
+@[simp] lemma subAt_append : subAt (subAt T x) y = subAt T (x ++ y) := by ext; simp
+@[gcongr] lemma subAt_mono {S T : tree A} (h : S ≤ T) : subAt S x ≤ subAt T x :=
+  Set.preimage_mono h
+
+/-- A variant of `List.drop` that takes values in `subAt` -/
+@[simps] def drop {T : tree A} (n : ℕ) (x : T) : subAt T (Tree.take n x).val :=
+  ⟨x.val.drop n, by simp⟩
+
+/-- Adjoint of `subAt` -/
```

Comment:
Please expand docstring with a more explicit description. ("Elements are prefixes of `x` or `x` with an element of `T` appended")